### PR TITLE
Upgrade terraform to 0.13.7

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/area control-plane
/kind enhancement

**What this PR does / why we need it**:
Upgrades terraform to 0.13.7. This is the first step of upgrading to 0.15 in order to fix gardener/gardener/issues/4838.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
* I ran successfully `make test-e2e` and `make docker-images` locally.
* I also ran successfully `make start` and `make start COMMAND=destroy` with a valid AWS configuration for one of my test shoots.
* For the provider validation status, see the comment below.

**Release note**:

```other operator
terraform has been upgraded to 0.13.7
```

### CODE REVIEW

**Comment by stoyanr** [comment]
Provider validation status:

AWS:
- [x] reconciliation and deletion of existing Infrastructure
- [x] creation and deletion of new Infrastructure (single zone, new vpc)
- [x] creation and deletion of new Infrastructure (single zone, existing vpc)
- [x] creation and deletion of new Infrastructure (multiple zones, new vpc)
- [x] creation and deletion of new Infrastructure (multiple zone, existing vpc)
- [ ] `gardener-extension-provider-aws` Infrastructure integration test (see gardener/gardener-extension-provider-aws/pull/439)

GCP:
- [x] reconciliation and deletion of existing Infrastructure
- [x] creation and deletion of new Infrastructure (single zone, new vpc)
- [x] creation and deletion of new Infrastructure (single zone, existing vpc)
- [ ] `gardener-extension-provider-gcp` Infrastructure integration test (see gardener/gardener-extension-provider-gcp/pull/345)

**Comment by ialidzhikov** [comment]
We had a sync with @stoyanr on how to proceed with the terraform version updates. We agreed to proceed first with the 0.13.7 update and to have a new terraformer minor release with it (`terraformer@v2.13.0`). When provider extensions successfully adopt `terraformer@v2.13.0`, we plan to proceed with the 0.14.11 and 0.15.5 updates in new minor releases - 0.14.11 update in `terraformer@v2.14.0` and 0.15.5 update `terraformer@2.15.0`. Meanwhile, if we need to do critical corrections, we will do it with patch releases of the corresponding terraformer minor release (for example provider-aws wants to update its terraformer-provider-aws - this can be performed with v2.13.1, v2.14.1 or v2.15.1 depending on which version of the terraformer is used currently by the provider-aws extesion).
We also agreed that https://github.com/gardener/gardener/issues/4978 is important for debugging and investigation purposes and it will be good if release of provider extension adopts terraform version bump together with a fix for https://github.com/gardener/gardener/issues/4978.

Last but not least, many thanks to @stoyanr for his work on this topic.
I will need some time to get familiar with the details and to run some basic tests also on my side.

/cc @rfranzke @dkistner @kon-angelo

File: test/utils/objects.go [hunk 1]
```diff
@@ -70,7 +70,7 @@ func PrepareTestObjects(ctx context.Context, c client.Client, namespacePrefix st
 	o.StateConfigMap = &corev1.ConfigMap{
 		ObjectMeta: metav1.ObjectMeta{Name: "tf-state", Namespace: o.Namespace},
 		Data: map[string]string{
-			StateKey: `some state`,
+			StateKey: `{"terraform_version":"0.13.7"}`,
```
Comment (ialidzhikov) [inline test/utils/objects.go:73]: Would it be possible to add a test case for 0.12.x and to verify that expected `state replace-provider` cmds are called?

**Review by ialidzhikov — COMMENTED** [review]
Looks good to me in general. I managed to test the following scenarios with gardener-extension-provider-aws and they worked well:
- reconciliation of existing Infrastructure
- deletion of existing Infrastructure
- creation and deletion of new Infrastructure (single zone, new vpc)
- creation and deletion of new Infrastructure (multiple zones, new vpc)
- creation and deletion of new Infrastructure (single zone, existing vpc)
- creation and deletion of new Infrastructure (multiple zone, existing vpc)

File: .gitignore [hunk 2]
```diff
@@ -3,6 +3,7 @@
 /tmp
 /local
 /dev
+/vendor/github.com/gardener/gardener/dev/
```
[the remark this answers is not in the document]
Comment (stoyanr) [reply .gitignore:6]: This directory actually showed up (several times) after running tests locally. I can remove it, but I am afraid we may end up committing  stuff to the repo that should not be there.

File: pkg/terraformer/terraformer.go [hunk 3]
```diff
@@ -50,6 +58,21 @@ var (
...
+	// to their sources (used with terraform 0.13 or higher)
+	providers = map[string]string{
+		"aws":         "hashicorp/aws",
+		"azurerm":     "hashicorp/azurerm",
+		"google":      "hashicorp/google",
+		"google-beta": "hashicorp/google-beta",
+		"openstack":   "terraform-provider-openstack/openstack",
+		"alicloud":    "hashicorp/alicloud",
+		"template":    "hashicorp/template",
+		"null":        "hashicorp/null",
```
[the remark this answers is not in the document]
Comment (stoyanr) [reply pkg/terraformer/terraformer.go:72]: I followed the ordering in https://github.com/gardener/terraformer/blob/master/build/all/terraform-bundle.hcl (and the same ordering for a subset of these entries are used in other `terraform-bundle.hcl` files as well). Not sure if an alphabetical order should be preferred over this one, as it's not being used anywhere else.

File: test/utils/objects.go (hunk 1 above)
Comment (stoyanr) [reply test/utils/objects.go]: Done

**Comment by stoyanr** [comment]
@ialidzhikov I added the tests as requested and replied to your other comments, PTAL.

**Review by ialidzhikov — APPROVED** [review]
/lgtm

**Pull request merged**